src/simplify.jl

```diff
@@ -79,3 +84,117 @@ function full_reduction(zxg::ZXGraph)
 
     return zxg
 end
+
+function full_reduction(bir::BlockIR)
+    zxd = convert_to_zxd(bir)
+    zxg = full_reduction(zxd)
+    chain = circuit_extraction(zxg)
+    return BlockIR(bir.parent, bir.nqubits, chain)
+end
+
+function compose_permutation(p1::Dict{Int, Int}, p2::Dict{Int, Int})
+    p2 = copy(p2)
+    p = Dict{Int, Int}()
+    for (k1, v1) in p1
+        if haskey(p2, v1)
+            p[k1] = p2[v1]
+            delete!(p2, v1)
+        else
+            p[k1] = v1
+        end
+    end
+    for (k2, v2) in p2
+        p[k2] = v2
+    end
+    return p
+end
+
+map_locations(d::Dict{T, T}, g::Gate) where {T <: Integer} = 
+    Gate(g.operation, map_locations(d, g.locations))
+map_locations(d::Dict{T, T}, cg::Ctrl) where {T <: Integer} = 
+    Ctrl(map_locations(d, cg.gate), map_locations(d, cg.ctrl))
+map_locations(d::Dict{T, T}, locs::Locations) where {T <: Integer} =
+    Locations(Tuple(haskey(d, i) ? d[i] : i for i in locs.storage))
+map_locations(d::Dict{T, T}, locs::CtrlLocations) where {T <: Integer} =
+    CtrlLocations(map_locations(d, locs.storage))
+
+function simplify_swap!(qc::Chain; replace_swap::Bool = true)
+    chain_after_swap = Chain()
+    loc_map = Dict{Int, Int}()
+    while length(qc.args) > 0
+        g = pop!(qc.args)
+        if g isa Gate
+            if g.operation === SWAP
+                loc1, loc2 = g.locations.storage[1:2]
+                loc_map = compose_permutation(Dict{Int, Int}(loc1 => loc2, loc2 => loc1), loc_map)
+                continue
+            end
+        end
+        pushfirst!(chain_after_swap.args, map_locations(loc_map, g))
```

Comment:
I feel SWAP should be simplified by looking at qubit dependency etc. but let's prob go with this simple one first.